### PR TITLE
Use implicit oauth flow

### PR DESCRIPTION
**Pull request by abemassry**
Tested locally, right now backwards compatible to use code, once the oauth server supports the implicit grant flow we can change the `response_type='code'` to `response_type='token'`

### CODE REVIEW

File: src/layouts/OAuth.js [hunk 1]
```diff
@@ -7,6 +7,22 @@ import { rawFetch } from '~/fetch';
...
+
+export function setSession(oauthToken = '', scopes = '') {
+  return (dispatch) => {
+    // Set these two so we can grab them on subsequent page loads
+    setStorage('authentication/oauth-token', oauthToken);
+    setStorage('authentication/scopes', scopes);
+    // Add all to state for this (page load) session
+    dispatch(setToken(oauthToken, scopes));
+  };
+}
```
Comment (eatonphil) [inline src/layouts/OAuth.js]: You now need to use `session.start()` here instead of this function. (I refactored this code a while ago into only the session module.)

Comment (abemassry) [inline src/layouts/OAuth.js]: so this would look something like `session.start(oauthToken, scopes)` or `dispatch(session.start(oauthToken, scopes))`

Comment (eatonphil) [inline src/layouts/OAuth.js]: dispatch(session.start()) but you also need to include the expires date as the final argument.

File: src/session.js [hunk 2]
```diff
@@ -3,22 +3,38 @@ import { APP_ROOT, LOGIN_ROOT } from '~/constants';
...
+}
+
 export function loginAuthorizePath(returnTo) {
+  const nonce = uuid4();
+  setStorage('authentication/nonce', nonce);
   /* eslint-disable prefer-template */
   return `${LOGIN_ROOT}/oauth/authorize?` +
          `client_id=${clientId}` +
          '&scopes=*' +
          '&response_type=code' + // TODO: response_type should be 'token' for implicit grant flow
```
Comment (abemassry) [inline src/session.js]: This can be merged right now but will not use the implicit flow till this is changed to `'&response_type=token'` that is waiting on OAuth server support

File: src/session.js [hunk 3]
```diff
@@ -3,22 +3,38 @@ import { APP_ROOT, LOGIN_ROOT } from '~/constants';
...
+         // `&state=${nonce}` + // TODO: response_type=token can optionally send a state
```
Comment (abemassry) [inline src/session.js]: For implicit flow this line would be uncommented

File: src/layouts/OAuth.js (hunk 1 above)
Comment (eatonphil) [inline src/layouts/OAuth.js]: This should still be removed.

File: src/layouts/OAuth.js [hunk 4]
```diff
@@ -7,6 +7,21 @@ import { rawFetch } from '~/fetch';
...
 import * as session from '~/session';
 
+import { getStorage, setStorage } from '~/storage';
+
+function getImplicitParams() {
+  const hashParams = window.location.hash.substr(1).split('&');
+  const params = {};
+  hashParams.forEach(function (hashParam) {
+    const hashParamParts = hashParam.split('=');
+    if (hashParam.match(/return/)) {
```
Comment (eatonphil) [inline src/layouts/OAuth.js]: Why are you regexing this? Can you just do `hashParam === 'return'`?

Comment (abemassry) [inline src/layouts/OAuth.js]: it returns something like return=http://localhost:3000/oauth/callback?return=/linodes

Comment (eatonphil) [inline src/layouts/OAuth.js]: Then why don't you do `hashParam.startsWith('return=')`?

Comment (abemassry) [inline src/layouts/OAuth.js]: I could do `hashParamParts[0] === 'return'` if you'd like

Comment (eatonphil) [inline src/layouts/OAuth.js]: Yep that's what I was looking for.

File: src/session.js [hunk 5]
```diff
@@ -3,22 +3,38 @@ import { APP_ROOT, LOGIN_ROOT } from '~/constants';
...
   window.location = location;
 }
 
+// Source: https://gist.github.com/jed/982883
+function uuid4(a) {
+  return a ? (
+    a ^ Math.random() * 16 >> a / 4
+  ).toString(16) : (
+    [1e7] + -1e3 + -4e3 + -8e3 + -1e11
+  ).replace(/[018]/g, uuid4);
```
Comment (eatonphil) [inline src/session.js]: This will call uuid4 on every match and insert the result.

File: src/session.js (hunk 3 above)
Comment (displague) [inline src/session.js]: I think we should enable state here and make this PR contingent on that support being present.  Even if this gets merged first, state could be abled here and left optional as it is in src/layouts/OAuth.js:85.

Comment (abemassry) [inline src/session.js]: I feel like that is outside of the scope of this PR and the code flow will be replaced once support exists on the OAuth server

File: src/session.js [hunk 6]
```diff
@@ -3,22 +3,39 @@ import { APP_ROOT, LOGIN_ROOT } from '~/constants';
...
-
 const AUTH_TOKEN = 'authentication/oauth-token';
 const AUTH_SCOPES = 'authentication/scopes';
 const AUTH_EXPIRES = 'authentication/expires';
 
+export function initializeAuthentication(dispatch) {
+  const token = getStorage('authentication/oauth-token') || null;
+  const scopes = getStorage('authentication/scopes') || null;
+  dispatch({ type: setToken, token, scopes });
+}
```
Comment (eatonphil) [inline src/session.js]: This is a duplicate of `initialize` below. Please use that instead.

Comment (eatonphil) [inline src/session.js]: This appears not to be used anymore. Must be a rebase product. Can safely be deleted.

File: src/layouts/OAuth.js [hunk 7]
```diff
@@ -7,12 +7,26 @@ import { rawFetch } from '~/fetch';
...
 import * as session from '~/session';
 
+import { getStorage, setStorage } from '~/storage';
+
+function getImplicitParams() {
+  const hashParams = window.location.hash.substr(1).split('&');
+  const params = {};
+  hashParams.forEach(function (hashParam) {
+    const hashParamParts = hashParam.split('=');
+    if (hashParamParts[0] === 'return') {
```
Comment (eatonphil) [inline src/layouts/OAuth.js]: See the comment in code (now) below. We don't need to decodeURIComponent because we don't encodeURIComponent on the oauth server.
